File: python/213_house-robber-ii.py
# Dynamic Programming Solution with Based on House Robber 1
class Solution:
    def rob(self, nums: List[int]) -> int:

        def find_max(nums):
            prev_max = 0
            cur_max = 0

            for n in nums:
                tmp = cur_max
                cur_max = max(cur_max, prev_max + n)
                prev_max = tmp

            return cur_max

        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]

        return max(find_max(nums[:-1]), find_max(nums[1:]))

File: go/213_house-robber-ii.go
func rob(nums []int) int {
	l := len(nums)
	if l == 0 {
		return 0
	}
	if l == 1 {
		return nums[0]
	}
	return max(helper(nums[:(l-1)]), helper(nums[1:]))
}

func helper(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	dp := []int{nums[0], max(nums[0], nums[1])}
	for i := 2; i < len(nums); i++ {
		index := i % 2
		dp[index] = max(dp[index]+nums[i], dp[(index+1)%2])
	}
	return max(dp[0], dp[1])
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
